### PR TITLE
[SPARK-28341][SQL] create a public API for V2SessionCatalog

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

The `V2SessionCatalog` has 2 functionalities:
1. work as an adapter: provide v2 APIs and translate calls to the `SessionCatalog`.
2. allow users to extend it, so that they can add hooks to apply custom logic before calling methods of the builtin catalog (session catalog).

To leverage the second functionality, users must extend `V2SessionCatalog` which is an internal class. There is no doc to explain this usage.

This PR does 2 things:
1. refine the document of the config `spark.sql.catalog.session`.
2. add a public abstract class `CatalogExtension` for users to write implementations.

TODOs for followup PRs:
1. discuss if we should allow users to completely overwrite the v2 session catalog with a new one.
2. discuss to change the name of session catalog, so that it's less likely to conflict with existing namespace names.

## How was this patch tested?

existing tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala [hunk 1]
```diff
@@ -62,23 +62,13 @@ class V2SessionCatalogSuite
...
   private def newCatalog(): TableCatalog = {
-    val newCatalog = new V2SessionCatalog(spark.sessionState)
+    val newCatalog = new V2SessionCatalog(spark.sessionState.catalog, spark.sessionState.conf)
     newCatalog.initialize("test", CaseInsensitiveStringMap.empty())
     newCatalog
   }
 
   private val testIdent = Identifier.of(Array("db"), "test_table")
 
-  test("Catalogs can load the catalog") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala]: v2 session catalog won't be load by name, after we remove the config.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -31,14 +31,20 @@ import org.apache.spark.sql.types.{LongType, StringType, StructType}
 
 class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAndAfter {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: Now this end-to-end test checks the actual behavior the end-users will hit, as we don't set a fake session catalog.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 3]
```diff
@@ -137,13 +144,12 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
     checkAnswer(spark.internalCreateDataFrame(rdd2, table.schema), Seq.empty)
   }
 
-  test("CreateTable: use default catalog for v2 sources when default catalog is set") {
-    val sparkSession = spark.newSession()
-    sparkSession.conf.set("spark.sql.catalog.testcat", classOf[TestInMemoryTableCatalog].getName)
-    sparkSession.conf.set("spark.sql.default.catalog", "testcat")
-    sparkSession.sql(s"CREATE TABLE table_name (id bigint, data string) USING foo")
+  test("CreateTable: use default catalog when default catalog is set") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: change the test name a little bit, because we will always use default catalog when the config is set, no matter what the table provider is.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 4]
```diff
@@ -174,26 +180,17 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
...
-    assert(table.name == "session.table_name")
-    assert(table.partitioning.isEmpty)
-    assert(table.properties == Map("provider" -> orc2).asJava)
-    assert(table.schema == new StructType()
-        .add("id", LongType, nullable = false)
-        .add("data", StringType))
-
-    val rdd = spark.sparkContext.parallelize(table.asInstanceOf[InMemoryTable].rows)
-    checkAnswer(spark.internalCreateDataFrame(rdd, table.schema), spark.table("source"))
+    // TODO: support write table from v2 session catalog.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: Note that, this is never supported. The test can pass before because we set a fake session catalog which returns writable tables.

**Comment by cloud-fan** [comment]
cc @rdblue @brkyvz @jose-torres @jzhuge

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  /**
+   * Tracks all the registered catalogs.
+   */
+  private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+
+  /**
+   * Looks up a catalog by name.
+   */
+  def getCatalog(name: String): CatalogPlugin = synchronized {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: Why use the Java-like "get"? I think `def catalog(name: String)` is sufficient.

I try to avoid adding `get` to names. It doesn't help understanding unless the method is a getter -- and this isn't -- so why include an extra word?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    catalogs.getOrElseUpdate(name, Catalogs.load(name, conf))
+  }
+
+  /**
+   * Returns the default catalog specified by config.
+   */
+  def getDefaultCatalog(): Option[CatalogPlugin] = {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: Same here. No need for `get`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: To clarify, this doesn't work because the ORC source is broken.

The test case is still valid, so I don't see a reason why we should get rid of it. I also think that it is a valid use to override the v2 session catalog implementation, which this relies on for tests.

**Comment by rdblue** [comment]
This looks okay overall, but it breaks the test for v2 providers with the session catalog. I don't think this can be committed until that is resolved. @cloud-fan, can you test whether one of the test providers will work in place of ORC v2 in this test?

The second issue is that this removes the ability to override the v2 session catalog implementation. I think we want to support that so you can modify the behavior of the v2 session catalog, but continue to use the v1 catalog for v1 cases. @cloud-fan, do you have any idea about how to make this possible?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala [hunk 7]
```diff
@@ -42,31 +37,15 @@ trait LookupCatalog extends Logging {
...
-  def sessionCatalog: Option[CatalogPlugin] = {
-    try {
-      Some(lookupCatalog(SESSION_CATALOG_NAME))
-    } catch {
-      case NonFatal(e) =>
-        logError("Cannot load v2 session catalog", e)
-        None
-    }
-  }
+  def sessionCatalog: TableCatalog = catalogManager.v2SessionCatalog
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala]: The session catalog needs to be configurable. This is how custom data sources / table formats will plugin.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Tracks all the registered catalogs.
+   */
+  private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+
+  /**
+   * Looks up a catalog by name.
+   */
+  def getCatalog(name: String): CatalogPlugin = synchronized {
+    catalogs.getOrElseUpdate(name, Catalogs.load(name, conf))
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: Since the nice error message handling is gone, if there's an error, what does it look like?

**Review by brkyvz — CHANGES_REQUESTED** [review]
@rdblue I still have some confusion around the `default.catalog` conf and `catalog.session`.

 - `spark.sql.default.catalog`: `Name of the default v2 catalog, used when a catalog is not identified in queries`
 - `spark.sql.catalog.session`: `Name of the default v2 catalog, used when a catalog is not identified in queries`.

The description of both is identical. The default v2 catalog can be configured on a session basis just like the `v2_session_catalog`. Why can't `DEFAULT_V2_CATALOG` be used to configure the session catalog?

**Comment by rdblue** [comment]
@brkyvz, the description must have been copied by mistake. The session property controls what implementation we use for the v2 session catalog. See my comment above for the difference between the v2 session catalog and the default catalog.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: the error message handling is in `defaultCatalog()` and `v2SessionCatalog`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          None
+      }
+    }
+  }
+
+  /**
+   * Returns the v2 session catalog specified by config.
+   */
+  lazy val v2SessionCatalog: TableCatalog = {
+    val implClsName = conf.getConf(StaticSQLConf.V2_SESSION_CATALOG_IMPL)
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: Why does this need to be a StaticConf? Wouldn't this make it impossible to plug in your own session catalog when using the Spark Shell for example?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/RequiresSessionCatalog.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2;
+
+import org.apache.spark.annotation.Unstable;
+import org.apache.spark.sql.catalyst.catalog.SessionCatalog;
+
+/**
+ * A marker interface to take the Spark internal session catalog, so that the catalog implementation
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/RequiresSessionCatalog.java]: This has a method, so it isn't a marker interface.

This should note that this is for implementations that want to use Spark's session catalog for storage, which is needed for writing replacements for V2SessionCatalog. Maybe also mention that this will go away when the SessionCatalog is removed.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/RequiresSessionCatalog.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * can hook in some custom logic before Spark calling the session catalog.
+ */
+@Unstable
+public interface RequiresSessionCatalog {
+  void initWithSessionCatalog(SessionCatalog catalog);
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/RequiresSessionCatalog.java]: If this doesn't replace the existing catalog initialization method, then it should have a different name. How about `setSessionCatalog`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    try {
+      loadV2SessionCatalog(implClsName)
+    } catch {
+      case NonFatal(e) =>
+        logError(s"Cannot create v2 session catalog: $implClsName", e)
+        loadV2SessionCatalog(StaticSQLConf.V2_SESSION_CATALOG_IMPL.defaultValueString)
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: This should not fall back to using an implementation other than the one the user configured. If the configured v2 session catalog cannot be instantiated, then the error should be thrown each time there is an attempt to use it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+   * Returns the default catalog specified by config.
+   */
+  // TODO(cloud-fan): This needs to be a def instead of val, because the default catalog config is
+  // a session config which can change at runtime. Shall we make it a static config instead?
+  def defaultCatalog(): Option[CatalogPlugin] = {
+    conf.defaultV2Catalog.flatMap { catalogName =>
+      try {
+        Some(catalog(catalogName))
+      } catch {
+        case NonFatal(e) =>
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: This should not swallow all non-fatal errors. It should catch CatalogNotFoundException only.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala [hunk 14]
```diff
@@ -42,31 +37,15 @@ trait LookupCatalog extends Logging {
...
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/LookupCatalog.scala]: Why is this `TableCatalog` and not `CatalogPlugin`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala (hunk 9 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: I think it makes sense to make this a static conf because the v2 session catalog cannot change after a session loads it the first time. Static confs can still be set using `--conf` right?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalogSuite.scala]: With your recent changes, I think the session catalog is loaded by name now, so this test should be restored.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 15]
```diff
@@ -176,26 +182,31 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
   }
 
   test("CreateTableAsSelect: use v2 plan and session catalog when provider is v2") {
-    spark.sql(s"CREATE TABLE table_name USING $orc2 AS SELECT id, data FROM source")
-
-    val testCatalog = spark.catalog("session").asTableCatalog
-    val table = testCatalog.loadTable(Identifier.of(Array(), "table_name"))
-
+    // TODO: `V2SessionCatalog#loadTable` should return writable table.
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: This should reference a JIRA issue to track the fix.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 16]
```diff
@@ -176,26 +182,31 @@ class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAn
...
+    val sessionCatalog = getV2SessionCatalog(newSession)
+    val table = sessionCatalog.loadTable(Identifier.of(Array(), "table_name"))
     assert(table.name == "session.table_name")
     assert(table.partitioning.isEmpty)
     assert(table.properties == Map("provider" -> orc2).asJava)
-    assert(table.schema == new StructType()
-        .add("id", LongType, nullable = false)
-        .add("data", StringType))
-
+    assert(table.schema == new StructType().add("a", IntegerType, nullable = false))
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: This should use the original test case to minimize changes.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 17]
```diff
@@ -22,25 +22,31 @@ import scala.collection.JavaConverters._
...
-  import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
-
   private val orc2 = classOf[OrcDataSourceV2].getName
 
+  private def getTestCatalog() = {
+    spark.sessionState.analyzer.catalogManager.catalog("testcat")
+      .asInstanceOf[TestInMemoryTableCatalog]
+  }
+
+  private def getV2SessionCatalog(session: SparkSession) = {
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: It is odd that this requires passing in the session when `getTestCatalog` does not. I would prefer not using this function in places where the session is different.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 18]
```diff
@@ -19,7 +19,7 @@ package org.apache.spark.sql.internal
 import org.apache.spark.SparkConf
 import org.apache.spark.annotation.{Experimental, Unstable}
 import org.apache.spark.sql.{ExperimentalMethods, SparkSession, UDFRegistration, _}
-import org.apache.spark.sql.catalog.v2.CatalogPlugin
+import org.apache.spark.sql.catalog.v2.{CatalogManager, CatalogPlugin}
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: Is this import used? Same with V2SessionCatalog. And check HiveSessionStateBuilder, too.

**Comment by cloud-fan** [comment]
I've updated the PR description to be more specific about what I'm fixing here, please take a look @rdblue @brkyvz

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 19]
```diff
@@ -111,6 +112,35 @@ class Analyzer(
 
   override protected def defaultCatalogName: Option[String] = conf.defaultV2Catalog
 
+  private lazy val v2SessionCatalog = new V2SessionCatalog(catalog, conf)
```
Comment (keypointt) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: minor and not a propose to change, just a quick check that the naming `V2SessionCatalog` or `SessionCatalogV2` ? 

I find V2 at the end like `WriteToDataSourceV2` or `CatalogV2Implicits` or `DataSourceV2Relation` in current code space, but also sometimes `V2SessionCatalog`, so I guess either way is ok?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 20]
```diff
@@ -39,20 +39,18 @@ import org.apache.spark.sql.util.CaseInsensitiveStringMap
 /**
  * A [[TableCatalog]] that translates calls to the v1 SessionCatalog.
  */
-class V2SessionCatalog(sessionState: SessionState) extends TableCatalog {
+class V2SessionCatalog(catalog: SessionCatalog, conf: SQLConf) extends TableCatalog {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala]: This change is necessary. We create `V2SessionCatalog` in `BaseSessionStateBuilder`, where there is no `SessionState` instance.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 21]
```diff
@@ -1914,9 +1914,12 @@ object SQLConf {
     .createOptional
 
   val V2_SESSION_CATALOG = buildConf("spark.sql.catalog.session")
-      .doc("Name of the default v2 catalog, used when a catalog is not identified in queries")
+      .doc("A catalog implementation which will be used as the session catalog for Spark " +
+        "queries. The implementation must either extend `CatalogExtension`, or implement " +
+        "`CatalogPlugin`. When extending `CatalogExtension`, the default session catalog " +
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: The class needs to implement `CatalogPlugin` to be loaded. I would rephrase this to "the implementation may extend `CatalogExtension` to be passed the default session catalog so that . . .".

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 22]
```diff
@@ -1914,9 +1914,12 @@ object SQLConf {
     .createOptional
 
   val V2_SESSION_CATALOG = buildConf("spark.sql.catalog.session")
-      .doc("Name of the default v2 catalog, used when a catalog is not identified in queries")
+      .doc("A catalog implementation which will be used as the session catalog for Spark " +
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: After the last sync discussion, I'm not sure that "will be used as the session catalog" is clear. I like the wording you used in the description of `CatalogExtension` when you referred to "Spark's built-in session catalog". How about changing this to "a catalog implementation that will be used in place of the Spark built-in session catalog for v2 operations".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 23]
```diff
@@ -39,20 +39,18 @@ import org.apache.spark.sql.util.CaseInsensitiveStringMap
...
+
+  def this(sessionState: SessionState) = this(sessionState.catalog, sessionState.conf)
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala]: Is this still used? Should we convert callers to use the new constructor instead?

**Comment by rdblue** [comment]
@cloud-fan, I had a couple of minor comments, but not blockers and overall it looks good to me. +1 when tests pass.

I like the idea of using the `TableCatalog` API for this more as I think about it. Good idea!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala (hunk 23 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala]: It's mostly for
```
def this() = this(SparkSession.active.sessionState)
```
Due to the limitation of Scala class constructor, I need to call `SparkSession.active.sessionState` twice if we remove the constructor here.

**Comment by cloud-fan** [comment]
@rdblue thanks for your review! I've addressed your comments and please take another look when you have time. Thanks!

**Comment by cloud-fan** [comment]
retest it please

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 24]
```diff
@@ -27,12 +27,18 @@ import org.apache.spark.sql.internal.SQLConf
...
-class CatalogManager(conf: SQLConf) extends Logging {
+class CatalogManager(conf: SQLConf, sessionCatalog: TableCatalog) extends Logging {
 
   private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
 
   def catalog(name: String): CatalogPlugin = synchronized {
-    catalogs.getOrElseUpdate(name, Catalogs.load(name, conf))
+    if (name.equalsIgnoreCase(CatalogManager.SESSION_CATALOG_NAME)) {
+      v2SessionCatalog.getOrElse {
+        throw new IllegalStateException("v2 session catalog not available")
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: why not `sessionCatalog`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 25]
```diff
@@ -27,12 +27,18 @@ import org.apache.spark.sql.internal.SQLConf
  * A thread-safe manager for [[CatalogPlugin]]s. It tracks all the registered catalogs, and allow
  * the caller to look up a catalog by name.
  */
-class CatalogManager(conf: SQLConf) extends Logging {
+class CatalogManager(conf: SQLConf, sessionCatalog: TableCatalog) extends Logging {
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: The flow in CatalogManager now seems a lot more confusing to me.
We now have a "built-in" session catalog, but none of the v2SessionCatalog methods (in an obvious manner) return that catalog.

Is the `sessionCatalog` in the Analyzer even match the `catalogManager.v2SessionCatalog` now, now that the configuration is optional?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 26]
```diff
@@ -101,15 +119,21 @@ object AnalysisContext {
  */
 class Analyzer(
     catalog: SessionCatalog,
+    v2SessionCatalog: TableCatalog,
     conf: SQLConf,
     maxIterations: Int)
   extends RuleExecutor[LogicalPlan] with CheckAnalysis with LookupCatalog {
 
+  // Only for tests.
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I don't link this too much. Do we need to change too much if we need to remove this constructor? It's bad to inject test related stuff in production code

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala (hunk 23 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala]: Where is the no-arg constructor called? I don't see it in master.

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala]: I don't think the no-arg constructor is needed after this commit. When this class was created by the catalog loader, it needed to have a no-arg constructor. After this commit, this class is only instantiated internally by Spark. We can remove all of these constructors.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 27]
```diff
@@ -164,7 +166,7 @@ abstract class BaseSessionStateBuilder(
    *
    * Note: this depends on the `conf` and `catalog` fields.
    */
-  protected def analyzer: Analyzer = new Analyzer(catalog, conf) {
+  protected def analyzer: Analyzer = new Analyzer(catalog, v2SessionCatalog, conf) {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: The `HiveSessionStateBuilder` must be updated as well because it overrides this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 26 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: There are more than 10 test suites using this constructor. BTW this is an existing constructor, this PR just add a comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 28]
```diff
@@ -27,12 +27,16 @@ import org.apache.spark.sql.internal.SQLConf
  * A thread-safe manager for [[CatalogPlugin]]s. It tracks all the registered catalogs, and allow
  * the caller to look up a catalog by name.
  */
-class CatalogManager(conf: SQLConf) extends Logging {
+class CatalogManager(conf: SQLConf, defaultSessionCatalog: TableCatalog) extends Logging {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: @brkyvz I agree with you that the logic was a bit confusing, so I refine it a little bit:
1. when the `V2_SESSION_CATALOG` is not set, return the default session catalog `V2SessionCatalog`.
2. when the `V2_SESSION_CATALOG` is set, try to instantiate it, and return the default session catalog if we hit problems during instantiating.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala [hunk 29]
```diff
@@ -47,13 +51,26 @@ class CatalogManager(conf: SQLConf) extends Logging {
...
+  }
+
+  // If the V2_SESSION_CATALOG config is specified, we try to instantiate the user-specified v2
+  // session catalog. Otherwise, return the default session catalog.
+  def v2SessionCatalog: CatalogPlugin = {
+    if (conf.getConf(SQLConf.V2_SESSION_CATALOG).isDefined) {
+      try {
+        catalogs.getOrElseUpdate(CatalogManager.SESSION_CATALOG_NAME, loadV2SessionCatalog())
+      } catch {
+        case NonFatal(_) => defaultSessionCatalog
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala]: I'd log the error. The user asked for a specific catalog, but we're giving them the default. There's no way to figure out the discrepancy without having the error

**Review by brkyvz — APPROVED** [review]
LGTM except the one case around logging the error. Should we also make V2SessionCatalog private[spark] after this case?

**Comment by brkyvz** [comment]
Another question is, some guidance on how a StagingTableCatalog can leverage the CatalogExtension API.

**Comment by cloud-fan** [comment]
If `V2SessionCatalog` implements `StagingTableCatalog` one day, it should still call `createTable`, `dropTable`, etc. somewhere which can be extended by `CatalogExtension`.

If users want to override the staging logic, they can implement `StagingTableCatalog` themselves.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by cloud-fan** [comment]
BTW, `V2SessionCatalog` is in the `execution` package, which is meant to be private, like the `catalyst` package.

**Comment by rdblue** [comment]
@cloud-fan, if execution is intended to be private, should we move those classes into catalyst?

**Comment by cloud-fan** [comment]
We can do it if `V2SessionCatalog` doesn't refer to anything in sql/core.
